runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/ApplianceWorkerHarness.java

```diff
@@ -0,0 +1,572 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Streaming Appliance worker harness implementation. */
+@Internal
+@ThreadSafe
+public final class ApplianceWorkerHarness implements StreamingWorkerHarness {
+  private static final Logger LOG = LoggerFactory.getLogger(ApplianceWorkerHarness.class);
+  private static final String MEMORY_MONITOR_EXECUTOR = "MemoryMonitor";
+  private static final String GET_WORK_EXECUTOR = "GetWorkDispatcher";
+  private static final String REFRESH_WORK_EXECUTOR = "RefreshWork";
+
+  private final DataflowWorkerHarnessOptions options;
+  private final long clientId;
+  private final WindmillServerStub windmillServer;
+  private final MetricTrackingWindmillServerStub getDataClient;
+  private final MemoryMonitor memoryMonitor;
+  private final ComputationStateCache computationCache;
+  private final StreamingWorkScheduler streamingWorkScheduler;
+  private final WorkCommitter workCommitter;
+  private final StreamingWorkerStatusReporter workerStatusReporter;
+  private final StreamingStatusPages statusPages;
+  private final DataflowExecutionStateSampler sampler;
+  private final ActiveWorkRefresher activeWorkRefresher;
+  private final BoundedQueueExecutor workExecutor;
+  private final ExecutorService memoryMonitoryExecutor;
+  private final ExecutorService getWorkExecutor;
+  private final AtomicBoolean isRunning;
+
+  private ApplianceWorkerHarness(
+      DataflowWorkerHarnessOptions options,
+      long clientId,
+      WindmillServerStub windmillServer,
+      MetricTrackingWindmillServerStub getDataClient,
+      MemoryMonitor memoryMonitor,
+      ComputationStateCache computationCache,
+      StreamingWorkScheduler streamingWorkScheduler,
+      WorkCommitter workCommitter,
+      StreamingWorkerStatusReporter workerStatusReporter,
+      StreamingStatusPages statusPages,
+      DataflowExecutionStateSampler sampler,
+      ActiveWorkRefresher activeWorkRefresher,
+      ExecutorService memoryMonitoryExecutor,
+      ExecutorService getWorkExecutor,
+      AtomicBoolean isRunning,
+      BoundedQueueExecutor workExecutor) {
+    this.options = options;
+    this.clientId = clientId;
+    this.windmillServer = windmillServer;
+    this.getDataClient = getDataClient;
+    this.memoryMonitor = memoryMonitor;
+    this.computationCache = computationCache;
+    this.streamingWorkScheduler = streamingWorkScheduler;
+    this.workCommitter = workCommitter;
+    this.workerStatusReporter = workerStatusReporter;
+    this.statusPages = statusPages;
+    this.sampler = sampler;
+    this.activeWorkRefresher = activeWorkRefresher;
+    this.memoryMonitoryExecutor = memoryMonitoryExecutor;
+    this.getWorkExecutor = getWorkExecutor;
+    this.isRunning = isRunning;
+    this.workExecutor = workExecutor;
+    // Register standard file systems.
+    FileSystems.setDefaultPipelineOptions(options);
+  }
+
+  public static ApplianceWorkerHarness fromOptions(DataflowWorkerHarnessOptions options) {
```

Comment:
At a high-level the difference between appliance/dispatcher/direct paths seem to be limited to:
- how to get the config
- how to get/schedule WorkProcessingContext (which then knows how to fetch/commit/heartbeat to right worker).

I think that moving things out of StreamingWorkerHarness into classes can help testing/readability but that there is still a lot of duplicated setup between the different harnesses which will make it harder to maintain.



Can we instead have a shared StreamingDataflowWorker that uses the objects and which just minimal injection differences are made when constructing for appliance/fanout/direct?

Reply:
sgtm, will redo some of the components to manage that

Reply:
Thanks! if you see ways to break up this change that would be great too.  It's really big and github reviews are particularly painful.  If we can just pull out various things from StreamingDataflowWorker to their own files/tests before getting to the direct path stuff I think it will help reviewing.  I will try to stay on top of the reviews so hopefully we don't get bogged down too much due to more PRs.

Reply:
sgtm!